### PR TITLE
Update WASI-enabled runtime to support running ASP.NET Core

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
As discussed with @lewing, @pavelsavara, and @radical this week.

It turns out that the WASI-enabled Mono build was already very close to being able to run the real ASP.NET Core. The missing pieces were wiring up about 10 more p/invoke calls and working around an issue with ALC.

For anyone else trying to understand the context here, this is **not** production-level support, but rather is moving forwards the prototype so we can continue to understand a wider range of scenarios.

The changes here should have zero impact on any build that doesn't have the `HOST_WASI` flag.

### CODE REVIEW

**Review by pavelsavara — APPROVED** [review]

File: src/mono/mono/metadata/appdomain.c [hunk 1]
```diff
@@ -774,6 +774,12 @@ ves_icall_System_Reflection_Assembly_InternalLoad (MonoStringHandle name_handle,
 	MonoAssembly *requesting_assembly = mono_runtime_get_caller_from_stack_mark (stack_mark);
 	MonoAssemblyLoadContext *alc = (MonoAssemblyLoadContext *)load_Context;
 
+#if HOST_WASI
+	// On WASI, mono_assembly_get_alc isn't supported
```
Comment (vargaz) [inline src/mono/mono/metadata/appdomain.c:778]: Is this a wasi limitation, or a general problem ?

File: src/mono/CMakeLists.txt [hunk 2]
```diff
@@ -505,9 +505,11 @@ if(GCC)
   set(WERROR "-Werror=return-type")
   set(WERROR_C "-Werror=implicit-function-declaration")
 
-  check_c_compiler_flag("-Werror=incompatible-pointer-types" WERROR_INCOMPATIBLE_POINTER_TYPES)
-  if(WERROR_INCOMPATIBLE_POINTER_TYPES)
-    set(WERROR_C "${WERROR_C} -Werror=incompatible-pointer-types")
+  if(NOT HOST_WASI)
+    check_c_compiler_flag("-Werror=incompatible-pointer-types" WERROR_INCOMPATIBLE_POINTER_TYPES)
```
Comment (vargaz) [inline src/mono/CMakeLists.txt:509]: Why is this needed ?

File: src/mono/mono/utils/mono-log-common.c [hunk 3]
```diff
@@ -110,7 +110,17 @@ mono_log_write_logfile (const char *log_domain, GLogLevelFlags level, mono_bool
 
 #ifdef HAVE_LOCALTIME_R
 		struct tm tod;
+#ifdef HOST_WASI
+		// In the WASI prototype, this is being compiled using WASI SDK which says that time() returns int64, but is
```
Comment (vargaz) [inline src/mono/mono/utils/mono-log-common.c:114]: Shouldn't libSystem.Native also be built using wasi ? Building some stuff using emscripten looks very fragile.

Comment (SteveSandersonMS) [reply src/mono/mono/utils/mono-log-common.c:114]: Yes, it definitely should. That's also what the TODO comment on L117 of this file says :)

File: src/mono/mono/metadata/appdomain.c (hunk 1 above)
Comment (SteveSandersonMS) [reply src/mono/mono/metadata/appdomain.c]: I've updated the comment here to clarify that it's not a permanent limitation but rather is something that remains to be done.

File: src/mono/mono/metadata/appdomain.c [hunk 4]
```diff
@@ -774,6 +774,12 @@ ves_icall_System_Reflection_Assembly_InternalLoad (MonoStringHandle name_handle,
 	MonoAssembly *requesting_assembly = mono_runtime_get_caller_from_stack_mark (stack_mark);
 	MonoAssemblyLoadContext *alc = (MonoAssemblyLoadContext *)load_Context;
 
+#if HOST_WASI
+	// On WASI, mono_assembly_get_alc isn't yet supported. However it should be possible to make it work.
+	if (!alc)
+		alc = mono_alc_get_default ();
+#endif
```
Comment (lambdageek) [inline src/mono/mono/metadata/appdomain.c:781]: @SteveSandersonMS Do you have a tracking issue for WASI support?  Might be good to start a task list of these kind of missing bits.

File: src/mono/mono/metadata/appdomain.c (hunk 1 above)
Comment (lambdageek) [reply src/mono/mono/metadata/appdomain.c]: > `mono_image_get_alc` fails with `Error: wasm trap: out of bounds memory access.`

That's not great.  Probably either `image` is null or it's junk.  Which means the assembly passed to `mono_assembly_get_alc` is null or junk.

So we're masking some underlying issue.   Does the trap happen early in startup or once the app is running?  can you build the driver with logging turned on (setenv `MONO_LOG_MASK=asm`, `MONO_LOG_LEVEL=debug`) - not sure if that will give any clues, but it might give an idea of when it's happening.

File: src/mono/CMakeLists.txt (hunk 2 above)
Comment (SteveSandersonMS) [reply src/mono/CMakeLists.txt]: When building under WASI it's being stricter about discarding `const` qualifiers. Existing code elsewhere (e.g., mono-rand.c:315), which I'm trying not to change in non-WASI builds, results in an error about that.

I've updated the config now so it's more precise about adding `no-incompatible-pointer-types-discards-qualifiers` for the WASI build rather than skipping `incompatible-pointer-types` entirely, so that should be more robust at least.

File: src/mono/mono/metadata/appdomain.c (hunk 4 above)
Comment (vargaz) [reply src/mono/mono/metadata/appdomain.c:781]: Is there a stack trace for when this happens ? Maybe the embedding app doesn't setup something correctly ?

File: src/mono/mono/metadata/appdomain.c (hunk 1 above)
Comment (SteveSandersonMS) [reply src/mono/mono/metadata/appdomain.c]: It happens during startup when trying to load the application's entrypoint assembly. It doesn't happen for a basic console app but does for an ASP.NET Core one, so it's something triggered inside ASP.NET Core, perhaps as it's trying to do some reflection over the entrypoint assembly.

The issue is that `mono_runtime_get_caller_from_stack_mark` on line 774 above is returning `NULL`, and so the attempt to dereference `assm->image` inside `mono_assembly_get_alc` will fail. I suppose internally it's something about stack walking being incomplete.

I'll add this to a list of things needing to be tracked down for complete WASI support in the future.

File: src/mono/mono/metadata/appdomain.c (hunk 4 above)
Comment (ShreyasJejurkar) [reply src/mono/mono/metadata/appdomain.c:781]: Exactly. I was looking for it. Didn't find. The issue will be easier to track PRs in one place and the community can discuss and learn as well along with these changes!

Comment (SteveSandersonMS) [reply src/mono/mono/metadata/appdomain.c:781]: Great idea - I've filed https://github.com/dotnet/runtime/issues/65895. There will probably be a bunch more things to add there over time.

Comment (SteveSandersonMS) [reply src/mono/mono/metadata/appdomain.c:781]: > Is there a stack trace for when this happens ?

Yes:

```
Error: wasm trap: out of bounds memory access
wasm backtrace:
    0: 0xc851 - <unknown>!mono_image_get_alc
    1: 0xc80b - <unknown>!mono_assembly_get_alc
    2: 0xc481 - <unknown>!ves_icall_System_Reflection_Assembly_InternalLoad
    3: 0xa55ba - <unknown>!ves_icall_System_Reflection_Assembly_InternalLoad_raw
    4: 0x3e1df6 - <unknown>!do_icall
    5: 0x3de329 - <unknown>!do_icall_wrapper
    6: 0x389273 - <unknown>!interp_exec_method
    7: 0x385119 - <unknown>!interp_runtime_invoke
    8: 0x30b64a - <unknown>!mono_jit_runtime_invoke
    9: 0x12a6ad - <unknown>!do_runtime_invoke
   10: 0x12c396 - <unknown>!mono_runtime_try_invoke
   11: 0x135eb5 - <unknown>!mono_runtime_invoke
   12: 0x40ea84 - <unknown>!mono_wasm_invoke_method
   13: 0x379b - <unknown>!start_dotnet
   14: 0x37e1 - <unknown>!__original_main
   15: 0x3622 - <unknown>!_start
```

More info here though: https://github.com/dotnet/runtime/pull/65888#discussion_r814829756

Comment (ShreyasJejurkar) [reply src/mono/mono/metadata/appdomain.c:781]: Yes @SteveSandersonMS, that's great. Maybe later you can put some links there about WASI stuff so that people can visit and learn more, as this is something totally new! 

Maybe this might be a TOC of issue. 
1. Heading as this is experimental 
2. WASI small intro in 2 lines. 
3. WASI official docs
4. List of efforts that the team (maybe you just at the moment) is planning/experimenting with. Link's to PR as well, if it's done!  
List of obstacles as of now (from .NET perspective, I heard one of them is ECMAScriptan that we are using as of now!) 
Sample projects with people can test out if they wanna!

Comment (vargaz) [reply src/mono/mono/metadata/appdomain.c:781]: Will investigate that, stack walks should definitely work with the interpreter even on wasm.

**Comment by SteveSandersonMS** [comment]
@lambdageek @vargaz I know this was already signed off by @pavelsavara but am unsure if either of you are looking for any further response to your comments before this is merged. If you could confirm you're happy I'll go ahead and merge it. There isn't really any risk here as this PR should not affect non-WASI builds and this is still in prototype phase. (cc @lewing)

**Comment by lambdageek** [comment]
@SteveSandersonMS :shipit:
